src/stackwalk.c

```diff
@@ -485,9 +529,28 @@ JL_DLLEXPORT jl_value_t *jl_lookup_code_address(void *ip, int skipC)
     return rs;
 }
 
-//for looking up functions from gdb:
-JL_DLLEXPORT void jl_gdblookup(uintptr_t ip)
+void jl_safe_print_codeloc(const char* func_name, const char* file_name,
+                           int line, int inlined) JL_NOTSAFEPOINT
 {
+    const char *inlined_str = inlined ? " [inlined]" : "";
+    if (line != -1) {
+        jl_safe_printf("%s at %s:%d%s\n", func_name, file_name, line, inlined_str);
+    }
+    else {
+        jl_safe_printf("%s at %s (unknown line)%s\n", func_name, file_name, inlined_str);
+    }
+}
+
+// Print function, file and line containing native instruction pointer `ip` by
+// looking up debug info. Prints multiple such frames when `ip` points to
+// inlined code.
+//
+// The need for `do_offset` is somewhat lost in history but apparently gives
+// better C stack traces in some cases. See issue #15845
+void jl_print_native_codeloc(uintptr_t ip, int do_offset) JL_NOTSAFEPOINT
+{
+    if (do_offset)
+        ip -= 1;
```

Comment:
@vtjnash perhaps you could comment on this `-1` (inherited from #15845)? Is it the difference between return address vs caller address or something? What's the most consistent way to deal with it?

Reply:
AFAICT this is the standard way. (it's how libunwind does it)

When you have an instruction address from unwinding you need to look at the previous insstruction since what you have is the return address. (The return address is the only thing recorded in a function call).

One caveat is that on ARM, we actually need (ip & ~(uintptr_t)1) - 1 since the pointer also contain information about the instruction set used. (libunwind/libunwind#131 FWIW, -2 might work in that case too......)

Reply:
Thanks! So what's the most consistent way forward here? It seems to me that it would be better to normalize these addresses so that they point to the call instruction during the platform-specific stack walking code? Then we can remove this flag.

Reply:
I tried to the ARMv7 case in https://github.com/JuliaLang/julia/pull/33380#discussion_r327985327